packages/react/src/components/button.tsx

```diff
@@ -91,247 +75,119 @@ export function UploadButton<
-    $props.__internal_upload_progress ?? 0,
-  );
-  const [files, setFiles] = useState<File[]>([]);
-
-  const { startUpload, isUploading, routeConfig } = useUploadThing(
-    $props.endpoint,
-    {
-      signal: acRef.current.signal,
-      headers: $props.headers,
-      onClientUploadComplete: (res) => {
-        if (fileInputRef.current) {
-          fileInputRef.current.value = "";
-        }
-        setFiles([]);
-        void $props.onClientUploadComplete?.(res);
-        setUploadProgress(0);
-      },
-      onUploadProgress: (p) => {
-        setUploadProgress(p);
-        $props.onUploadProgress?.(p);
-      },
-      onUploadError: $props.onUploadError,
-      onUploadBegin: $props.onUploadBegin,
-      onBeforeUploadBegin: $props.onBeforeUploadBegin,
-    },
-  );
-
-  const uploadFiles = useCallback(
-    (files: File[]) => {
-      startUpload(files, fileRouteInput).catch((e) => {
-        if (e instanceof UploadAbortedError) {
-          void $props.onUploadAborted?.();
-        } else {
-          throw e;
-        }
-      });
-    },
-    [$props, startUpload, fileRouteInput],
-  );
-
-  const { fileTypes, multiple } = generatePermittedFileTypes(routeConfig);
-
-  const inputProps = useMemo(
-    () => ({
-      type: "file",
-      ref: fileInputRef,
-      multiple,
-      accept: generateMimeTypes(fileTypes).join(", "),
-      onChange: (e: React.ChangeEvent<HTMLInputElement>) => {
-        if (!e.target.files) return;
-        const selectedFiles = Array.from(e.target.files);
-
-        $props.onChange?.(selectedFiles);
-
-        if (mode === "manual") {
-          setFiles(selectedFiles);
-          return;
-        }
-
-        void uploadFiles(selectedFiles);
-      },
-      disabled: fileTypes.length === 0,
-      tabIndex: fileTypes.length === 0 ? -1 : 0,
-    }),
-    [$props, fileTypes, mode, multiple, uploadFiles],
-  );
-
-  if ($props.__internal_button_disabled) inputProps.disabled = true;
-  if ($props.disabled) inputProps.disabled = true;
-
-  const state = (() => {
-    if ($props.__internal_state) return $props.__internal_state;
-    if (inputProps.disabled) return "disabled";
-    if (!inputProps.disabled && !isUploading) return "ready";
-    return "uploading";
-  })();
-
-  usePaste((event) => {
-    if (!appendOnPaste) return;
-    if (document.activeElement !== fileInputRef.current) return;
+  const { className, content, appearance, ...rootProps } =
+    props as unknown as UploadButtonProps<TRouter, TEndpoint>;
 
-    const pastedFiles = getFilesFromClipboardEvent(event);
-    if (!pastedFiles) return;
+  const cn = rootProps.config?.cn ?? defaultClassListMerger;
 
-    let filesToUpload = pastedFiles;
-    setFiles((prev) => {
-      filesToUpload = [...prev, ...pastedFiles];
-
-      $props.onChange?.(filesToUpload);
-
-      return filesToUpload;
-    });
-
-    if (mode === "auto") void uploadFiles(files);
-  });
+  return (
+    <Primitive.Root<TRouter, TEndpoint> {...(rootProps as any)}>
```

Comment:
ooc, what's making it required for an any cast here?

Reply:
I'm not 100% sure what the exact issue is (the type error is too big to show in my editor) but it's related to the input field on `UploadthingComponentProps`
```ts
ExtendObjectIf<
    inferEndpointInput<TRouter[TEndpoint]>,
    {
      /**
       * The input to the endpoint, as defined using `.input()` on the FileRouter endpoint
       * @see https://docs.uploadthing.com/api-reference/server#input
       */
      input: inferEndpointInput<TRouter[TEndpoint]>;
    }
  >;
```